### ISSUE TITLE
Compile all unit tests into a single executable

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#9155]
Our builds take a huge amount of space due to amount of test executables and due to how big each of them are, as they each statically link against the entire scylla code-base. This has lead us to build even debug builds without debug symbols, to prevent  storage requirements getting out of hands.

Since df822e09e we run test cases in parallel, so in theory we could compile all unit tests into the same executable and would still e able to run them in parallel. This would surely require some mechanical transformation of each test, most likely sprinkling anonymous namespaces here-and-there and handling some tougher cases of duplicated symbols, it doesn't sound too hard.

I have proposed this in the past but was told that this direction is not worth exploring because we should just go for modules instead. While I partially agree, there are two reasons while I still think this is worth exploring:
* Even though we are on C++20 already, modules are not production ready in either Clang, nor GCC. In Clang they outright don't work at all (produced binary crashes), and while GCC's modules are more promising, we can't switch back due to some coroutine related bugs.
* Many tests genuinely need to include most of the scylla core code and modules won't help here.

The hope is that doing this would drastically reduce the storage requirements, and also cut down the link time, which while not the largest contributor to build time, it is a significant one, especially in debug mode.

**Comment by elcallio** [comment]
Compiling a libscylla.a and statically linking to it, instead of *.o should have much the same effect, and still allow builing a single unit test when needed.

**Comment by denesb** [comment]
> Compiling a libscylla.a and statically linking to it, instead of *.o should have much the same effect, and still allow builing a single unit test when needed.

Do you mean that if we are not using `-Wl,--whole-archive`, then linking will eliminate symbols unused by the unit test? This is actually something that would be easy to try and use without any downsides (that I see).

**Comment by avikivity** [comment]
Also, many tests can reduce their dependencies (although it will take a monumental effort).

**Comment by nyh** [comment]
> Compiling a libscylla.a and statically linking to it, instead of *.o should have much the same effect, and still allow builing a single unit test when needed.

It's also possible for the build system to have rules for both individual tests executables and the one-big-test-executable. Just the one-big-test-executable gets linked in the default build, but you can ask specifically to build one test executable. Similar to how we those "*_g" targets in our build file, which don't get built by default but as soon as you want one, you can ask to build it.

**Comment by bhalevy** [comment]
@tchaikov with building scylla-core as a shared library this won't be needed, right?
Is there an issue opened for that?

**Comment by denesb** [comment]
We have https://github.com/scylladb/scylladb/issues/4142.

**Comment by tchaikov** [comment]
ahh, i need to restart my work on modularized seastar. but shared library and C++ module are orthogonal. we can have both of them or either of them.

**Comment by bhalevy** [comment]
> ahh, i need to restart my work on modularized seastar. but shared library and C++ module are orthogonal. we can have both of them or either of them.

True.  But I still hope they will be aligned (i.e. if a library consists a c++ module, it should be fully contained in it)

**Comment by denesb** [comment]
I approached the shared-library from an angle of build-time, which did not improve much, so I abandoned the effort.
I did however see huge reduction in space requirements.

**Comment by yaronkaikov** [comment]
Do we have someone assign to this?

**Comment by nyh** [comment]
I should never have closed https://github.com/scylladb/scylladb/issues/5 which asked to solve this problem 9 years (!) ago.

**Comment by nyh** [comment]
> Compiling a libscylla.a and statically linking to it, instead of *.o should have much the same effect, and still allow builing a single unit test when needed.

I agree with @elcallio that his approach (static library) will surely reduce the size of test executables _somewhat_. But it can still leave plenty of duplication because many test executables will still contain duplicates of general code used by many tests (Seastar infrastructure, cql environment, etc.) - although some duplicate would indeed be avoided. How much duplication is left vs avoided, I don't know - maybe it's worth testing.

The duplication can be avoided **completely** by using a _shared_ library (as I suggested in #5). This will slow down the run, but we should use this idea _only_  for the debug build, which is much bigger than the other builds.
But I do agree with @denesb that building a single test _executable_ instead of a shared library, is a reasonable alternative to a shared library, because in both approaches you need to re-link the entire thing (into a single executable or single shared library) on every source code change. The downside I see with a single executable is that I think it will be harder to achieve properly (we still need that single executable to be able to run the different tests, and we have multiple types and variants of test frameworks used by different tests).

**Comment by denesb** [comment]
Note that for the shared library approach, you need to compile all source files twice: once with `-fpic` (for tests) and once without it (for the scylla executable).

> The downside I see with a single executable is that I think it will be harder to achieve properly (we still need that single executable to be able to run the different tests, and we have multiple types and variants of test frameworks used by different tests).

We use the same boost test framework for all our unit tests. I don't see any problem with compiling all these into a single executable, beyond some mechanical problems like having to make sure symbols don't collide (see the opening comment).

**Comment by nyh** [comment]
> Note that for the shared library approach, you need to compile all source files twice: once with `-fpic` (for tests) and once without it (for the scylla executable).

Not if for the problematic build mode (namely, debug) we only build with fpic.

> We use the same boost test framework for all our unit tests. I don't see any problem with compiling all these into a single executable, beyond some mechanical problems like having to make sure symbols don't collide (see the opening comment).

I thought we also had "non-boost" tests in the "test/unit" directory. Maybe it's time we got rid of them, or maybe they are few enough that we don't care what their size is.

Anyway, I'm not against your solution of a single executable.

**Comment by yaronkaikov** [comment]
> I should never have closed #5 which asked to solve this problem 9 years (!) ago.

I agree ;-)

**Comment by yaronkaikov** [comment]
@mykaul This is the issue we talked about, hope we can somehow move forward with this,

**Comment by mykaul** [comment]
> ahh, i need to restart my work on modularized seastar. but shared library and C++ module are orthogonal. we can have both of them or either of them.

@tchaikov - where does that stand?

**Comment by tchaikov** [comment]
@mykaul hi Yaniv, 

- C++20 module support already landed. see https://github.com/scylladb/seastar/pull/1605 .
- build seastar as a shared library also finished. see https://github.com/scylladb/seastar/pull/1484

the modularized work is now tracked by https://github.com/scylladb/scylladb/issues/12962 . i am still working on https://github.com/scylladb/scylladb/issues/2717

**Comment by yaronkaikov** [comment]
@mykaul This one can really help us to avoid those spot termination, splitting those tests between multi machines with shorten test time

**Comment by nyh** [comment]
@tchaikov why is C++20 module support needed, or relevant, to this issue?
Unless I'm missing something, you don't need modules to link all the tests into a single executable, or to create a shared library and link that into all the tests, or the other ideas floated here or in #5.

**Comment by tchaikov** [comment]
> @tchaikov why is C++20 module support needed, or relevant, to this issue? 

hmm, i must be missing something. i didn't claim that C++20 module is relevant to this issue. on the contrary, i think it's not related. see https://github.com/scylladb/scylladb/issues/9155#issuecomment-1494256966 .

> Unless I'm missing something, you don't need modules to link all the tests into a single executable, or to create a shared library and link that into all the tests, or the other ideas floated here or in #5.

**Comment by syuu1228** [comment]
I'm testing this for a while, here are ideas how to implement this:


Here's idea how to make C++ unit tests into single binary.


First, modify Boost test unit files to link single binary.
To do so, we need to add new .cc file (I named it combined_test.cc) just for main function, and it should contain following lines:
```
#define BOOST_TEST_MODULE "module name"
#include <boost/test/included/unit_test.hpp>
```
Then we need to add `BOOST_AUTO_TEST_SUITE()` and `BOOST_AUTO_TEST_SUITE_END()` for each unit test file to define each unit test file as test suite.
And also we need to remove `#define BOOST_TEST_MODULE` from unit test file since we don't want to define main function on these files.

It will be some thing like this:
```
diff --git a/test/boost/enum_option_test.cc b/test/boost/enum_option_test.cc
index 933204f48c..2e19d45ed7 100644
--- a/test/boost/enum_option_test.cc
+++ b/test/boost/enum_option_test.cc
@@ -6,7 +6,7 @@
  * SPDX-License-Identifier: AGPL-3.0-or-later
  */

-#define BOOST_TEST_MODULE core
+//#define BOOST_TEST_MODULE core

 #include <boost/test/unit_test.hpp>

@@ -19,6 +19,8 @@

 #include "utils/enum_option.hh"

+BOOST_AUTO_TEST_SUITE(enum_option_test)
+
 namespace po = boost::program_options;

 namespace {
@@ -159,3 +161,5 @@ BOOST_AUTO_TEST_CASE(test_non_string) {
     BOOST_CHECK_EQUAL(format<numbers>(numbers::TWO), "2");
     BOOST_CHECK_EQUAL(format<numbers>(static_cast<numbers::enumeration>(77)), "?unknown");
 }
+
+BOOST_AUTO_TEST_SUITE_END()
```


Next, we also need to modify Seastar test unit files into single binary.
Seastar test is implemented on Boost test but there are number of difference, I found that we may not share same main function.
So "single binary" actually will be two single binaries, one for Boost test, another for Seastar test.

Just like Boost test we need to add new .cc file (I named it seastar_combined_test.cc) just for main function, and it should contain following lines:
```
#include "test/lib/scylla_test_case.hh"
```
Then we need to add `BOOST_AUTO_TEST_SUITE()` and `BOOST_AUTO_TEST_SUITE_END()` for each unit test file to define each unit test file as test suite.
And also we need to remove #include "test/lib/scylla_test_case.hh" since it provides main function, then we should add following headers:
```
#undef SEASTAR_TESTING_MAIN
#include <seastar/testing/test_case.hh>
```
After this, each unit files will not have main function.

It will be something like this:
```
diff --git a/test/boost/btree_test.cc b/test/boost/btree_test.cc
index b0bcd66dbc..6b352af651 100644
--- a/test/boost/btree_test.cc
+++ b/test/boost/btree_test.cc
@@ -9,11 +9,15 @@
 #include <boost/test/unit_test.hpp>
 #include <fmt/core.h>
 
-#include "test/lib/scylla_test_case.hh"
+#undef SEASTAR_TESTING_MAIN
+
+#include <seastar/testing/test_case.hh>
 #include <seastar/testing/thread_test_case.hh>
 #include "test/unit/tree_test_key.hh"
 #include "utils/intrusive_btree.hh"
 
+BOOST_AUTO_TEST_SUITE(btree_test)
+
 using namespace intrusive_b;
 using namespace seastar;
 
@@ -587,3 +591,5 @@ SEASTAR_TEST_CASE(test_exception_safety_of_clone_linear) {
 SEASTAR_TEST_CASE(test_exception_safety_of_clone_large) {
     return test_exception_safety_of_clone(2534);
 }
+
+BOOST_AUTO_TEST_SUITE_END()
```


Then we need to modify configure.py to link boost test unit files and combined_test.cc into single binary, and link seastar test unit files and seastar_combined_test.cc into single binary.
To do so we need to remove each units from `scylla_tests` and `pure_boost_tests` since we don't want to generate executable for each tests, and add combined_test and seastar_combined_test instead.
Also we need to define dependencies on deps[x], to add dependencies on combined_test and seastar_combined_test.


It will be something like this:
```
diff --git a/configure.py b/configure.py
index ee1aadaae2..b01538431b 100755
--- a/configure.py
+++ b/configure.py
@@ -430,6 +430,8 @@ modes = {
 }
 
 scylla_tests = set([
+    'test/boost/combined_test',
+    'test/boost/seastar_combined_test',
     'test/boost/UUID_test',
     'test/boost/pretty_printers_test',
     'test/boost/cdc_generation_test',
@@ -447,7 +449,6 @@ scylla_tests = set([
     'test/boost/bytes_ostream_test',
     'test/boost/cache_algorithm_test',
     'test/boost/cache_mutation_reader_test',
-    'test/boost/cached_file_test',
     'test/boost/caching_options_test',
     'test/boost/canonical_mutation_test',
     'test/boost/cartesian_product_test',
@@ -480,8 +481,6 @@ scylla_tests = set([
     'test/boost/dirty_memory_manager_test',
     'test/boost/duration_test',
     'test/boost/dynamic_bitset_test',
-    'test/boost/enum_option_test',
-    'test/boost/enum_set_test',
     'test/boost/extensions_test',
     'test/boost/error_injection_test',
     'test/boost/filtering_test',
@@ -582,7 +581,6 @@ scylla_tests = set([
     'test/boost/wasm_test',
     'test/boost/wasm_alloc_test',
     'test/boost/bptree_test',
-    'test/boost/btree_test',
     'test/boost/radix_tree_test',
     'test/boost/double_decker_test',
     'test/boost/stall_free_test',
@@ -1365,6 +1363,7 @@ deps = {
 }
 
 pure_boost_tests = set([
+    'test/boost/combined_test',
     'test/boost/anchorless_list_test',
     'test/boost/auth_passwords_test',
     'test/boost/auth_resource_test',
@@ -1378,8 +1377,6 @@ pure_boost_tests = set([
     'test/boost/crc_test',
     'test/boost/duration_test',
     'test/boost/dynamic_bitset_test',
-    'test/boost/enum_option_test',
-    'test/boost/enum_set_test',
     'test/boost/idl_test',
     'test/boost/json_test',
     'test/boost/keys_test',
@@ -1453,6 +1450,14 @@ for t in sorted(perf_tests):
     deps[t] = [t + '.cc'] + scylla_tests_dependencies + perf_tests_seastar_deps
     deps[t] += ['test/perf/perf.cc', 'seastar/tests/perf/linux_perf_event.cc']
 
+deps['test/boost/combined_test'] += [
+    'test/boost/enum_option_test.cc',
+    'test/boost/enum_set_test.cc',
+]
+deps['test/boost/seastar_combined_test'] += [
+    'test/boost/btree_test.cc',
+    'test/boost/cached_file_test.cc',
+]
 deps['test/boost/mutation_reader_test'] += ['test/lib/dummy_sharder.cc' ]
 deps['test/boost/multishard_combining_reader_as_mutation_source_test'] += ['test/lib/dummy_sharder.cc' ]
```


After building new test suceeded, combined_test will work like this:
```
./tools/toolchain/dbuild ./build/debug/test/boost/combined_test --list_content
enum_option_test*
    test_parsing*
    test_parsing_error*
    test_formatting*
    test_formatting_unknown*
    test_ordered_map*
    test_multiple_parse*
    test_multiple_format*
    test_non_string*
enum_set_test*
    enum_max_sequence*
    enum_min_sequence*
    enum_valid_sequence*
    set_contains*
    full_set*
    set_from_mask*
    set_enable*
    set_enable_if*
    set_remove*
    set_iterator*
    set_add*
    set_toggle*
    
./tools/toolchain/dbuild build/debug/test/boost/combined_test --run_test=enum_option_test
Running 8 test cases...

*** No errors detected
```

And seastar_combined_test too:
```
./tools/toolchain/dbuild build/debug/test/boost/seastar_combined_test --list_content
btree_test*
    test_ops_empty_tree*
    test_plain_key_pointer*
    test_double_insert*
    test_cookie_find*
    test_double_erase*
    test_remove_corner_case*
    test_end_iterator*
    test_next_to_end_iterator*
    test_clear*
    test_post_clear*
    test_iterator_erase*
    test_iterator_equal*
    test_find_all*
    test_lower_bound_all*
    test_upper_bound_all*
    test_insert_iterator_index*
    test_insert_before*
    test_iterators*
    test_data_self_iterator*
    test_singular_tree_ptr*
    test_range_erase*
    test_clone*
    test_insert_before_hint*
    test_tree_size*
    test_swap_between_trees_1*
    test_swap_between_trees*
    test_unlink_leftmost_without_rebalance*
    test_exception_safety_of_clone_linear*
    test_exception_safety_of_clone_large*
cached_file_test*
    test_file_wrapper*
    test_no_crash_on_dtor_after_oom*
    test_concurrent_population*
    test_reading_from_small_file*
    test_eviction_via_lru*
    test_invalidation*
    
./tools/toolchain/dbuild build/debug/test/boost/seastar_combined_test --run_test=btree_test
Running 29 test cases...
CHK 0
CHK 1
CHK 3
CHK 32
WARNING: debug mode. Not for benchmarking or production
INFO  2024-09-05 21:12:49,382 seastar - Reactor backend: linux-aio
INFO  2024-09-05 21:12:49,393 seastar - Perf-based stall detector creation failed (EACCESS), try setting /proc/sys/kernel/perf_event_paranoid to 1 or less to enable kernel backtraces: falling back to posix timer.
random-seed=2813299372
==1==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!

*** No errors detected
```


Then we also need to modify test.py to handle single binary unit test.
Here's what we need to do on test.py:
 - need to modify boost test list, since it will show up the single binary, but does not show up each unit tests since it because listing each test binaries.
   It can be change by modifying "unit_test_list" target defined on configure.py, since test.py using the result of `ninja unit_test_list`.
 - current --list_content parser breaks on the single binary, because it doesn't able to parse "suite" correctly.
   It need to fix.
 - A single test must be executable per file even all unit tests are linked into single binary.
   To explain the above example of "combined_test", when we run `./test.py boost/enum_option_test:test_parsing`, it should be call `./build/debug/test/boost/combined_test enum_option_test.test_parsing`.
   To implement it we need some hack on BoostTest.__init__.
   On BoostTest.__init__, shortname is the program name, and boost_args are parameters passed to the boost test executable.
   We need to replace them when we handle the single binary unit test.
   On "combined_test" case, shortname should be 'combined_test' and boost_args should be ['--run-test=enum_option_test.test_parsing'].
   We can detect the signle binary unit test by file existance check, if executable is not found it should be the single binary unit test.
   (Once we replaced all C++ test to the single binary test, we don't need to detect it)

**Comment by syuu1228** [comment]
> Does `BOOST_AUTO_TEST_SUITE(btree_test)` also introduce a namespace behind the scenes, eliminating the problem of global variable name clashes?

Verified, seems like `BOOST_AUTO_TEST_SUITE(btree_test)` defines `namespace btree_test`.

I checked boost-test header, the namespace is defined at `BOOST_AUTO_TEST_SUITE_WITH_DECOR`:
```
#define BOOST_AUTO_TEST_SUITE_WITH_DECOR( suite_name, decorators )      \
namespace suite_name {                                                  \
BOOST_AUTO_TU_REGISTRAR( suite_name )(                                  \
    BOOST_STRINGIZE( suite_name ),                                      \
    __FILE__, __LINE__,                                                 \
    decorators );                                                       \
/**/

#define BOOST_AUTO_TEST_SUITE_NO_DECOR( suite_name )                    \
    BOOST_AUTO_TEST_SUITE_WITH_DECOR(                                   \
        suite_name,                                                     \
        boost::unit_test::decorator::collector_t::instance() )          \
/**/

#if BOOST_PP_VARIADICS
#define BOOST_AUTO_TEST_SUITE( ... )                                    \
    BOOST_TEST_INVOKE_IF_N_ARGS( 1,                                     \
        BOOST_AUTO_TEST_SUITE_NO_DECOR,                                 \
        BOOST_AUTO_TEST_SUITE_WITH_DECOR,                               \
        __VA_ARGS__)                                                    \
/**/

#else /* BOOST_PP_VARIADICS */

#define BOOST_AUTO_TEST_SUITE( suite_name )                             \
    BOOST_AUTO_TEST_SUITE_NO_DECOR( suite_name )                        \
/**/


#endif /* BOOST_PP_VARIADICS */
```

When I did test above, `BOOST_AUTO_TEST_SUITE()` sometimes caused a compilation error depends on where I added.
It is probably because it changes namespace on the code.

**Comment by syuu1228** [comment]
Opened DRAFT PR which only contain C++ part: https://github.com/scylladb/scylladb/pull/21443
Still working on test.py to support single executable tests.

**Comment by roydahan** [comment]
How can we move this PR forward?

**Comment by avikivity** [comment]
> How can we move this PR forward?

Before we answer that, we need to understand if it should be moved forward.
